File: src/main/scala/datastructure/Util/ThreadUtils.java
package datastructure.Util;

import java.util.concurrent.*;
import java.util.concurrent.locks.LockSupport;

public class ThreadUtils {
    public static <T> Future<T> intoFuture(ExecutorService pool, Callable<T> callable) {
        try {
            return pool.submit(() -> {
                return callable.call();
            });
        } catch (Exception e) {
            throw new RuntimeException("Error executing in separate transaction", e);
        }
    }

    public final static ExecutorService DEFAULT = createDefaultPool();

    public static ExecutorService createDefaultPool() {
        int threads = Runtime.getRuntime().availableProcessors()*2;
        int queueSize = threads * 25;
        return new ThreadPoolExecutor(threads / 2, threads, 30L, TimeUnit.SECONDS, new ArrayBlockingQueue<>(queueSize),
                new CallerBlocksPolicy());
//                new ThreadPoolExecutor.CallerRunsPolicy());
    }

    static class CallerBlocksPolicy implements RejectedExecutionHandler {
        @Override
        public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
            if (!executor.isShutdown()) {
                // block caller for 100ns
                LockSupport.parkNanos(100);
                try {
                    // submit again
                    executor.submit(r).get();
                } catch (InterruptedException | ExecutionException e) {
                    throw new RuntimeException(e);
                }
            }
        }
    }
}

File: src/main/scala/datastructure/Util/ThreadUtils.scala
package datastructure.Util

import java.util.concurrent._
import java.util.concurrent.locks.LockSupport

object ThreadUtils {
  def intoFuture[T](pool: ExecutorService , callable: Callable[T]): Future[T] = {
    pool.submit(() => {
      def foo(): T = callable.call
      foo()
    })
  }
  def createDefaultPool :ThreadPoolExecutor = {
    val threads = Runtime.getRuntime.availableProcessors * 2
    val queueSize = threads * 25
    new ThreadPoolExecutor(threads / 2, threads, 30L, TimeUnit.SECONDS, new ArrayBlockingQueue[Runnable](queueSize), new CallerBlocksPolicy)
  }

  class CallerBlocksPolicy() extends RejectedExecutionHandler {
    override def rejectedExecution(r: Runnable, executor: ThreadPoolExecutor): Unit = {
      if (!executor.isShutdown) { // block caller for 100ns
        LockSupport.parkNanos(100)
        try // submit again
        executor.submit(r).get
        catch {
          case e@(_: InterruptedException | _: ExecutionException) =>
            throw new RuntimeException(e)
        }
      }
    }
  }
}
